tests/test_curvature.py

```diff
@@ -0,0 +1,187 @@
+                diag = [1.0, 2.0]
+            self.hessian = torch.diag(torch.tensor(diag)).to(DEVICE)
+
+        self.hessian = torch.nn.Parameter(self.hessian, requires_grad=False)
+
+    def forward(self, x):
+        return torch.dot(x, torch.matmul(self.hessian, x))
+
+
+def value_grad_hess(f, point, dtype):
+    value = f(point)
+    grad = torch.autograd.functional.jacobian(f, point)
+    hess = torch.autograd.functional.hessian(f, point)
+    return value.to(dtype), grad.to(dtype), hess.to(dtype)
+
+
+#def hyperboloid_eq(x, a, c):
+#    return (x[:-1] ** 2) / (a ** 2) - x[-1] ** 2 / c ** 2
+
+
+#@pytest.mark.parametrize('curve_func', ['moosavi', 'golden', 'lee_graph']) # the first one will fail
+@pytest.mark.parametrize('curve_func', ['golden', 'lee_graph'])
+def test_hyperboloid(curve_func):
+    a = 1
+    c = 3
+    f = lambda x: curve_utils.hyperboloid_graph(x[0], x[1], a=a, c=c)
+    x = 2.0
+    y = 2.0
+    point = torch.tensor([x, y]).to(DEVICE)
+    z, grad, hess = value_grad_hess(f, point, dtype=torch.double)
+
+    shape_operator, principal_curvatures, principal_directions = curve_utils.local_response_curvature_alternates(grad, hess, curve_func)
+
+    u = z.cpu().numpy() / c
+    analytic_gauss_curvature = -c**2 / (c**2 + (a**2 + c**2) * u**2)**2
+
+    np.testing.assert_allclose(
+        analytic_gauss_curvature,
+        torch.det(shape_operator).cpu().numpy(),
+        rtol=1e-6
+    )
+
+
+@pytest.mark.parametrize('dimensions', [2, 3, 5, 10])
+@pytest.mark.parametrize('radius', [0.1, 1, 2, 3, 10])
+#@pytest.mark.parametrize('curve_func', ['moosavi', 'lee_level']) # the first one will fail
+@pytest.mark.parametrize('curve_func', ['lee_level'])
+def test_sphere(dimensions, radius, curve_func):
+    f = QuadraticFunction(np.ones(dimensions)).to(DEVICE)
+    #point = torch.tensor(np.hstack((np.zeros(dimensions-1), [radius]))).to(DEVICE)
+    point = torch.tensor(np.ones(dimensions) / np.sqrt(dimensions) * radius).to(DEVICE)
+    value, pt_grad, pt_hess = value_grad_hess(f, point, dtype=torch.double)
+
+    iso_shape_operator, iso_curvatures, iso_directions = curve_utils.local_response_curvature_alternates(pt_grad, pt_hess, curve_func)
+    iso_curvatures = iso_curvatures.detach().cpu().numpy()
+
+    assert iso_curvatures.var() < 1e-16
+
+    expected_gaussian_curvature = 1 / (radius ** (dimensions - 1))
+    np.testing.assert_allclose(np.abs(np.prod(iso_curvatures)), expected_gaussian_curvature)
+
+
+@pytest.mark.parametrize('dimensions', [2, 3, 5])
+@pytest.mark.parametrize('radius', [0.1, 1, 2])
+def test_coordinate_transform_sphere(dimensions, radius):
+    f = QuadraticFunction(np.ones(dimensions)).to(DEVICE)
+    #point = torch.tensor(np.hstack((np.zeros(dimensions-1), [radius]))).to(DEVICE)
+    point = torch.tensor(np.ones(dimensions) / np.sqrt(dimensions) * radius).to(DEVICE)
+    
+    value, pt_grad, pt_hess = value_grad_hess(f, point, dtype=torch.double)
+
+    coordinate_transformations = [
```

Comment:
would be more elegant to do this via parametrization as well. Then you see directly which transformation fails.